src/runtime.rs

```diff
@@ -280,113 +247,151 @@ impl Runtime {
+
+        // Extract the changed_at and durability values from the top of the stack;
+        // these reflect the queries which were executed thus far and which
+        // led to the cycle.
+        let changed_at = from_stack.last().unwrap().changed_at;
+        let durability = from_stack.last().unwrap().durability;
+
+        // Identify the cycle participants:
+        let cycle = {
+            let mut v = vec![];
+            dg.for_each_cycle_participant(
+                from_id,
+                &mut from_stack,
                 database_key_index,
-                error.to,
-                query_stack.iter().map(|query| query.database_key_index),
-                &mut cycle,
+                to_id,
+                |aq| v.push(aq.database_key_index),
             );
-            cycle.push(database_key_index);
-
-            assert!(!cycle.is_empty());
-
-            for active_query in query_stack
-                .iter_mut()
-                .filter(|query| cycle.iter().any(|key| *key == query.database_key_index))
-            {
-                active_query.cycle = cycle.clone();
+            Cycle::new(Arc::new(v))
+        };
+        debug!("cycle {:?}", cycle.debug(db));
+
+        // Identify cycle recovery strategy:
+        let recovery_strategy = self.mutual_cycle_recovery_strategy(db, &cycle);
+        debug!("cycle recovery strategy {:?}", recovery_strategy);
+
+        // If using fallback, we have to mark the cycle participants, so they know to recover.
+        match recovery_strategy {
+            CycleRecoveryStrategy::Panic => {}
+            CycleRecoveryStrategy::Fallback => {
+                // Mark the cycle participants, so they know to recover:
+                dg.for_each_cycle_participant(
+                    from_id,
+                    &mut from_stack,
+                    database_key_index,
+                    to_id,
+                    |aq| {
+                        aq.cycle = Some(cycle.clone());
+                    },
+                );
             }
+        }
 
-            crate::CycleError {
+        self.local_state.restore_query_stack(from_stack);
+
+        CycleDetected {
+            recovery_strategy,
+            cycle_error: CycleError {
                 cycle,
                 changed_at,
-                durability: Durability::MAX,
-            }
+                durability,
+            },
         }
     }
 
-    pub(crate) fn mark_cycle_participants(&self, err: &CycleError<DatabaseKeyIndex>) {
-        for active_query in self
-            .local_state
-            .borrow_query_stack_mut()
-            .iter_mut()
-            .rev()
-            .take_while(|active_query| {
-                err.cycle
-                    .iter()
-                    .any(|e| *e == active_query.database_key_index)
-            })
+    fn mutual_cycle_recovery_strategy(
+        &self,
+        db: &dyn Database,
+        cycle: &Cycle,
+    ) -> CycleRecoveryStrategy {
```

Comment:
Do we envision other stategies beyond Panic/Fallback? If not, this can be written in a simpler way by just chceking if any element is a Panic.

Reply:
Yes, I do. =) I'm working towards "fixed point" to better integrate chalk and salsa.

Reply:
(that said, I do often also try to write my code "as if" more match variants will be added unless I am quite certain they won't)